### ISSUE TITLE
Separate labels from main script

### ISSUE THREAD

**Issue by jschwager** [Installomator/Installomator#1999]
Based on my daily use of Installomator, I believe the overall experience could be significantly enhanced by separating the labels from the main Installomator script. Here's a proposed approach on how to implement this process:

**Initial Check:** On its first run, Installomator should verify the existence of a specified local file path to determine if a separate file containing the labels case statement is present. Additionally, it should check for a version number within that file to ensure it's current.

**File Update:** If the labels file is either missing or outdated, Installomator should automatically download and update the local labels file to ensure it has the latest information.

**Execution:** During execution, Installomator should source the labels case statement directly from the local labels file. This will streamline the process and ensure that the script always utilizes the most up-to-date labels.

By implementing this logic, the management and updating of labels would become more efficient, improving the overall functionality and user experience of Installomator.

**Comment by TomBoehm** [comment]
I like the idea!
And if we had a  local **folder path** containing label files, we could apply additional local changes to override the Installomator default labels.

**Comment by acodega** [comment]
The file update step is the hardest part. How do we maintain the security of that update? Think about how we verify a pkg is signed and notarized. 

How do we verify the download is valid and verify the website isn't compromised? (Yes by verifying team ID we are assuming the developer account has not been compromised)

That's not to say I'm rejecting the idea I would love to have more conversation about this. I like how quickly a few of us have commented on this idea.

**Comment by gilburns** [comment]
I like the idea. The current Installomator repo becomes just the script and logic. The labels then become a separately maintained repo. 

The main Installomator repo could be updated as the script requires or new features are introduced. This would allow you to only see PRs against that part of the code. It can then be updated and a new signed pkg built as needed.

The labels become a separate repo that generates it's own signed pkg with an automated build on some frequency (daily, weekly or monthly) build cycle. I feel like that would help the Installomator team more quickly get updated or new labels approved and into a production release pkg. So instead of pulling the labels directly that can't be easily verified for authenticity, a pkg would be available that could be verified.

There are currently 250+ open PRs, many or which are labels. That really should be an easy and quick thing to get out the door. I feel like separating the labels would allow for an update cadence to get new labels available much quicker. (The cadence between Installomator 10.5 and 10.6 was almost a year. That's a very long time for labels to get into a production non-beta build.)

**Comment by jschwager** [comment]
My only other thought to this approach would be that there should be some functionality baked in to allow folks to use their own repo if desired for the label package, or some other method to allow for custom/fix labels to be added. In its current state it is actually very nice to be able to make quick adjustments to fix things on the fly if you know what you’re doing.

**Comment by TomBoehm** [comment]
@acodega You are right. The hardest part is to trust all those download links. To be notarized just states that there is no malware in the Installomator script itself.

**Comment by ReArmedHalo** [comment]
I think this would be a fantastic idea! We use installomator heavily and have had to create several "override" labels in the installomator script in Jamf to accommodate these.

One thing I think would be neat is if, when using your own repository for labels, you have the option of combining with the official repository. This would allow the best of both worlds in my opinion; allowing admins to have their own collection of custom labels they need for their organization + overrides for fixes that may not exist upstream yet, while not moving the full burden of keeping a downstream label repository updated constantly.

Maybe for update checks and signature verification/anti-tampering, the official repository could be verified by MD5 checksum checking which would also allow for update checks without needing a version string in a label file in the official repo.

**Comment by acodega** [comment]
@TomBoehm I'm not referring to trusting the download links inside of Installomator. What I mean is how do you verify that the labels you've downloaded is a complete list, and it hasn't been maliciously changed?

The challenge with using a checksum is where do we host the checksum value? If it's hosted in the same place as the labels, they can both be compromised at the same time.

With a signed/notarized pkg, both our GitHub *and* our Apple Developer account has to be compromised together for an attack to be successful. (Not saying this is bullet proof protection)

**Comment by gilburns** [comment]
Fleshing out my suggestions above a bit more...  The only thing that would ever be "live" downloaded would be the signed Installomator labels pkg. That can be verified with a signature like every other pkg file Installomator downloads. It would install into some known path that the main script knows how to find. Say something like /usr/local/bin/Installomator/labels/

Then administrators can distribute their own vetted "managed labels" with their own methods. (Jamf, Intune, etc...) This would be any title they want to install that has not been added to the "official" labels pkg release. Those labels would live in a path something like this: /usr/local/bin/Installomator/managed-labels/

The Installomator script then searches both locations for a label. I'd suggest that "managed" labels should take precedence over the "official" Installomator labels. Thus allowing for their own special customization and needs. This eliminates the entire security dilemma about live downloading potentially malicious code. Both sets of labels are either signed off on by the Installomator community, or the Mac admin themselves.

**Comment by acodega** [comment]
Okay I see what you mean. The admin distributes those managed labels themselves. Like via MDM. I was thinking of something similar the other day.

To be honest I think this is better served by an Installomator fork. Fork is too heavy of a phrase but maybe call it an Installomator bolt on.

One of our primary goals is to make Installomator independent. Meaning you have the script and that's all you need. Even if you don't have MDM.

Yes, you'll ask "What about the extra features for Datadog, swiftDialog, and others? That's not agnostic."

It comes down to support. Yes we're open source and can just tell people "Open a PR" but we (I, especially) feel a certain obligation to maintaining a product that is easy for us to support. We are careful about adding additional features because of the maintainer burden. Someone has to be around to help improve things. Sometimes it's other contributors but sometimes that's hard, literally inheriting someone else's code.

Long story short, I would recommend if folks want to develop this that they treat it as a fork first, run with it for some time, and then consider a PR.

**Comment by gilburns** [comment]
Maybe the above PR will never make it into the code, but I feel it has merit to consider it.

Even if you don't want to completely separate the labels from the script, I feel it still might be beneficial, longterm, for the Installomator team to consider a separate repo for the labels. This would allow a clear distinction between reviewing fairly simple label PR updates in one repo and then PR changes to the script code in the other main repo. The main repo could then add the script repo as a dependent project and the script could still embed the labels from the other repo.

My 2¢ 😁

**Comment by gilburns** [comment]
https://github.com/Installomator/Installomator/pull/2068

**Comment by abstertee** [comment]
I agree with all the above points and would love to see a new repo or just a separate file that can be downloaded like what has been mentioned.  Happy to help where I can.
